### PR TITLE
http: Track active requests and wait for last to finish

### PR DESCRIPTION
**Pull request by promag**
This PR calls `evhttp_free` before releasing event base. But according to `evhttp_free` [docs](https://github.com/libevent/libevent/blob/1cea01d6d5c84337dac663e5464059ccd2d6a8dd/include/event2/http.h#L208-L217):
> Works only if no requests are currently being served.

So this PR also tracks active requests with libevent and waits for last request to finish. This requires libevent 2.1 due to `evhttp_request_set_on_complete_cb` (https://github.com/libevent/libevent/blob/master/whatsnew-2.1.txt).

Finally, the call to `evhttp_free` is done in the event base loop to avoid concurrency issues.

Now test/functional/feature_abortnode.py quits normally, not due to socket timeouts.

### CODE REVIEW

**Comment by promag** [comment]
Minimal fix in #15363.

**Comment by promag** [comment]
@laanwj ping, again :sweat_smile:

**Comment by promag** [comment]
There is an issue which is fixed in #19434 - complete callback isn't called if client closes the connection.

**Comment by jonatack** [comment]
Interesting! -- this PR brings the run time of `test/functional/feature_abortnode.py` down from 32 seconds to only 2.

**Comment by jonatack** [comment]
Concept ACK, the code looks fine, tests pass and `feature_abortnode.py` for instance runs much faster.

I did see something odd which may be related to the issue you mention in https://github.com/bitcoin/bitcoin/pull/19420#issuecomment-653285641:
```
bcli: bitcoin-cli -rpcuser=... -rpcpassword=... getblockhash 637819 exited 1 (after 60 other errors) 'error: Could not connect to the server 127.0.0.1:8332

Make sure the bitcoind server is running and that you are connecting to the correct RPC port.
'; we have been retrying command for --bitcoin-retry-timeout=60 seconds; bitcoind setup or our --bitcoin-* configs broken?
keysend: pay: Lost connection to the RPC socket.
Lost connection to the RPC socket.
fundchannel: Lost connection to the RPC socket.
autoclean: Lost connection to the RPC socket.
```

Also, you may want to see https://github.com/libevent/libevent/issues/643 and https://github.com/libevent/libevent/pull/591.

**Review by jonatack — COMMENTED** [review]
Concept ACK, the code looks fine, tests pass and `feature_abortnode.py` for instance runs much faster.

Also, you may want to see https://github.com/libevent/libevent/issues/643 and https://github.com/libevent/libevent/pull/591.

**Comment by fjahr** [comment]
Concept ACK

**Comment by promag** [comment]
@laanwj don't you miss reviewing this code? 😅

**Comment by jonatack** [comment]
> So this PR also tracks active requests with libevent and waits for last request to finish. This requires libevent 2.1 due to `evhttp_request_set_on_complete_cb` ([libevent/libevent@`master`/whatsnew-2.1.txt](https://github.com/libevent/libevent/blob/master/whatsnew-2.1.txt)).

Does this change require bumping the minimum enforced libevent version from 2.0.21 to 2.1?

File: src/httpserver.cpp [hunk 1]
```diff
@@ -146,6 +146,10 @@ static std::unique_ptr<WorkQueue<HTTPClosure>> g_work_queue{nullptr};
 static std::vector<HTTPPathHandler> pathHandlers;
 //! Bound listening sockets
 static std::vector<evhttp_bound_socket *> boundSockets;
+//! Track active requests
+static Mutex g_requests_mutex;
+static std::condition_variable g_requests_cv;
```
Comment (jonatack) [inline src/httpserver.cpp:151]: 662cba4254b8f54d02f7d64717ae5a7e13b5a502 maybe add these headers
```diff
--- a/src/httpserver.cpp
+++ b/src/httpserver.cpp
@@ -21,8 +21,10 @@
 #include <util/threadnames.h>
 #include <util/translation.h>
 
+#include <condition_variable>
 #include <deque>
 #include <memory>
+#include <mutex>
 #include <stdio.h>
```

File: src/httpserver.cpp [hunk 2]
```diff
@@ -459,14 +475,22 @@ void StopHTTPServer()
...
     }
     boundSockets.clear();
+    {
+        WAIT_LOCK(g_requests_mutex, lock);
+        while (!g_requests.empty()) {
+            g_requests_cv.wait(lock);
+        }
+    }
+    if (eventHTTP) {
+        event_base_once(eventBase, -1, EV_TIMEOUT, [](evutil_socket_t, short, void*) {
```
Comment (jonatack) [inline src/httpserver.cpp:485]: ea724b9 curious why you added this line

The other place in this codebase that calls `event_base_once()` to schedule a one-time event is torcontrol.cpp in `InterruptTorControl`.

**Review by jonatack — COMMENTED** [review]
Light approach ACK ea724b9ca40c58881ad51058079641c1e5edb81b based on code review, looked at the libevent repo, doc/dependencies and configure.ac, rebased, debug build, ran unit tests, re-verified that `test/functional/feature_abortnode.py` runs far faster (under 2 seconds with debug build on a slow CPU), ran mainnet/signet/testnet nodes with http logging on and made a few CLI calls, checked nodes shutdown. Modulo we would need to bump the minimum required version of libevent from 2.0.21 to 2.1.

<details><summary>debug log on signet shutdown</summary><p>

```
2022-03-23T11:57:59Z [http] Received a POST request for / from 127.0.0.1:33432
2022-03-23T11:57:59Z [init] Interrupting HTTP server
2022-03-23T11:57:59Z [init] tor: Thread interrupt
2022-03-23T11:57:59Z [init] Shutdown: In progress...
2022-03-23T11:57:59Z [shutoff] Unregistering HTTP handler for / (exactmatch 1)
2022-03-23T11:57:59Z [shutoff] Unregistering HTTP handler for /wallet/ (exactmatch 0)
2022-03-23T11:57:59Z [shutoff] Stopping HTTP server
2022-03-23T11:57:59Z [shutoff] Waiting for HTTP worker threads to exit
2022-03-23T11:57:59Z [addcon] addcon thread exit
2022-03-23T11:57:59Z [torcontrol] RemoveLocal(xxx.onion:38333)
2022-03-23T11:57:59Z [torcontrol] torcontrol thread exit
2022-03-23T11:57:59Z [opencon] opencon thread exit
2022-03-23T11:57:59Z [http] Exited http event loop
2022-03-23T11:57:59Z [shutoff] Waiting for HTTP event thread to exit
2022-03-23T11:57:59Z [shutoff] Stopped HTTP server
2022-03-23T11:57:59Z [net] net thread exit
2022-03-23T11:57:59Z [msghand] msghand thread exit
2022-03-23T11:58:00Z [i2paccept] i2paccept thread exit
2022-03-23T11:58:00Z [shutoff] DumpAnchors: Flush 2 outbound block-relay-only peer addresses to anchors.dat started
2022-03-23T11:58:00Z [shutoff] DumpAnchors: Flush 2 outbound block-relay-only peer addresses to anchors.dat completed (0.01s)
2022-03-23T11:58:00Z [scheduler] scheduler thread exit
2022-03-23T11:58:00Z [shutoff] I2P: Destroying session 99c6a79e84
2022-03-23T11:58:00Z [shutoff] Writing 0 unbroadcast transactions to disk.
2022-03-23T11:58:00Z [shutoff] Dumped mempool: 7.4e-05s to copy, 0.008535s to dump
2022-03-23T11:58:00Z [shutoff] [default wallet] Releasing wallet
2022-03-23T11:58:00Z [shutoff] [blank] Releasing wallet
2022-03-23T11:58:00Z [shutoff] [encrypted blank descriptor] Releasing wallet
2022-03-23T11:58:00Z [shutoff] [disable private keys] Releasing wallet
2022-03-23T11:58:00Z [shutoff] [disable private keys and blank] Releasing wallet
2022-03-23T11:58:00Z [shutoff] [default settings] Releasing wallet
2022-03-23T11:58:00Z [shutoff] Shutdown: done
```
</p></details>

Commit message: "http: Relese server before waiting for event base loop exit" -> s/Relese/Release/

File: src/httpserver.cpp (hunk 2 above)
Comment (promag) [reply src/httpserver.cpp:485]: This is a way to call `evhttp_free` in the event base thread, meaning it's not called concurrently with requests.

**Comment by hebasto** [comment]
> Does this change require bumping the minimum enforced libevent version from 2.0.21 to 2.1?

Done in #24681.

**Comment by achow101** [comment]
Are you still working on this?

**Comment by achow101** [comment]
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.

**Comment by fjahr** [comment]
I have reopened this PR in #26742. Comments by @jonatack have been addressed.

**Pull request closed without merging**